# <div class='alert alert-success'>  Q&A Chat Bots</div>

**We are gonna work in some kind of story like Jane went to the store and Mike went to the bedroom**

So if we ask any question like is Mike in the balcony the asnwer should be **No**

In [1]:
import pickle as pi
import numpy as np

In [2]:
with open("train_qa.txt", "rb") as f: 
    traindata =  pi.load(f)

In [3]:
with open("test_qa.txt", "rb") as f:  
    testdata =  pi.load(f)

## **Lets explore the  data now**

In [4]:
type(traindata)

list

In [5]:
type(testdata)

list

In [6]:
len(traindata)

10000

In [7]:
len(testdata)

1000

In [8]:
traindata[3]

(['Mary',
  'moved',
  'to',
  'the',
  'bathroom',
  '.',
  'Sandra',
  'journeyed',
  'to',
  'the',
  'bedroom',
  '.',
  'Mary',
  'went',
  'back',
  'to',
  'the',
  'bedroom',
  '.',
  'Daniel',
  'went',
  'back',
  'to',
  'the',
  'hallway',
  '.',
  'Sandra',
  'went',
  'to',
  'the',
  'kitchen',
  '.',
  'Daniel',
  'went',
  'back',
  'to',
  'the',
  'bathroom',
  '.',
  'Daniel',
  'picked',
  'up',
  'the',
  'football',
  'there',
  '.',
  'Daniel',
  'went',
  'to',
  'the',
  'bedroom',
  '.'],
 ['Is', 'Daniel', 'in', 'the', 'bedroom', '?'],
 'yes')

**<font color='green'>So  after the story if it is asked "Is Daniel in the bedroom?" the answer should be yes</font>**

Its a tuple of list so if we want only the story itself then

In [9]:
' '.join(traindata[3][0])

'Mary moved to the bathroom . Sandra journeyed to the bedroom . Mary went back to the bedroom . Daniel went back to the hallway . Sandra went to the kitchen . Daniel went back to the bathroom . Daniel picked up the football there . Daniel went to the bedroom .'

**Lets check for another example**

This is the whole tuple

In [10]:
traindata[5]

(['Sandra',
  'got',
  'the',
  'football',
  'there',
  '.',
  'Mary',
  'went',
  'to',
  'the',
  'bedroom',
  '.'],
 ['Is', 'Mary', 'in', 'the', 'bedroom', '?'],
 'yes')

This is story

In [11]:
' '.join(traindata[5][0])

'Sandra got the football there . Mary went to the bedroom .'

This is the question

In [12]:
' '.join(traindata[5][1])

'Is Mary in the bedroom ?'

This is the answer

In [13]:
print(traindata[5][2])

yes


### Lets set up a vocabulary of our data set using all the training + test data

In [14]:
alll=traindata+testdata

In [15]:
len(alll)

11000

**We know `'set'` data type contains only unique values so we are going to check the all unique values we have in our data**

In [16]:
vocab=set() #initialised empty set
for story,question,answer in alll:
    vocab=vocab.union(set(story))
    vocab=vocab.union(set(question))
    

In [17]:
vocab.add('yes')

In [18]:
vocab.add('no')

In [19]:
vocab

{'.',
 '?',
 'Daniel',
 'Is',
 'John',
 'Mary',
 'Sandra',
 'apple',
 'back',
 'bathroom',
 'bedroom',
 'discarded',
 'down',
 'dropped',
 'football',
 'garden',
 'got',
 'grabbed',
 'hallway',
 'in',
 'journeyed',
 'kitchen',
 'left',
 'milk',
 'moved',
 'no',
 'office',
 'picked',
 'put',
 'the',
 'there',
 'to',
 'took',
 'travelled',
 'up',
 'went',
 'yes'}

**These are all the unique data we have**

In [20]:
len(vocab)

37

In [21]:
vocab_len=len(vocab)+1
#We are adding 1 because later we would be needing a place holder

In [22]:
vocab_len

38

**Lets check the length of the stories and later check which is the maximum length**

In [23]:
all_stories_length=[len(d[0]) for d in alll]

In [24]:
all_stories_length

[12,
 26,
 39,
 52,
 64,
 12,
 24,
 36,
 49,
 61,
 13,
 25,
 38,
 50,
 63,
 13,
 25,
 37,
 49,
 61,
 24,
 37,
 48,
 71,
 83,
 13,
 25,
 37,
 50,
 62,
 12,
 24,
 35,
 46,
 71,
 12,
 24,
 37,
 50,
 62,
 13,
 25,
 38,
 50,
 62,
 12,
 24,
 36,
 48,
 60,
 13,
 25,
 39,
 50,
 62,
 13,
 26,
 38,
 49,
 60,
 12,
 24,
 36,
 49,
 61,
 13,
 26,
 39,
 51,
 62,
 13,
 25,
 38,
 52,
 65,
 12,
 24,
 37,
 48,
 61,
 13,
 25,
 38,
 51,
 64,
 12,
 25,
 39,
 53,
 65,
 12,
 24,
 36,
 48,
 60,
 12,
 24,
 36,
 49,
 61,
 12,
 25,
 38,
 50,
 61,
 25,
 37,
 49,
 60,
 72,
 13,
 25,
 36,
 48,
 60,
 12,
 24,
 36,
 48,
 60,
 12,
 23,
 36,
 47,
 59,
 12,
 24,
 35,
 48,
 61,
 12,
 23,
 36,
 49,
 62,
 12,
 25,
 37,
 48,
 59,
 12,
 25,
 37,
 49,
 60,
 12,
 24,
 36,
 49,
 61,
 12,
 25,
 37,
 49,
 61,
 12,
 24,
 37,
 49,
 60,
 12,
 25,
 38,
 50,
 61,
 12,
 24,
 36,
 49,
 61,
 12,
 24,
 35,
 48,
 61,
 12,
 25,
 37,
 49,
 62,
 12,
 24,
 38,
 50,
 62,
 12,
 25,
 38,
 50,
 62,
 12,
 24,
 36,
 48,
 60,
 12,
 24,
 36,
 49,
 60,


**Max length is**

In [25]:
max_story_len=max(all_stories_length)

In [26]:
max_story_len

156

**Lets check the max question length**

In [27]:
max_ques_len=max([len(d[1]) for d in alll])

In [28]:
max_ques_len

6

## <font color='blue'>Now we are going to see how we are going to vectorise our data</font>

In [29]:
from keras_preprocessing.sequence import pad_sequences
from keras.preprocessing.text import Tokenizer

In [30]:
tok=Tokenizer(filters=[])
tok.fit_on_texts(vocab)

In [31]:
tok.word_index

{'moved': 1,
 '?': 2,
 'garden': 3,
 '.': 4,
 'sandra': 5,
 'daniel': 6,
 'the': 7,
 'grabbed': 8,
 'office': 9,
 'up': 10,
 'bathroom': 11,
 'to': 12,
 'john': 13,
 'put': 14,
 'milk': 15,
 'went': 16,
 'down': 17,
 'no': 18,
 'dropped': 19,
 'in': 20,
 'discarded': 21,
 'is': 22,
 'apple': 23,
 'back': 24,
 'journeyed': 25,
 'there': 26,
 'left': 27,
 'picked': 28,
 'got': 29,
 'bedroom': 30,
 'kitchen': 31,
 'took': 32,
 'hallway': 33,
 'mary': 34,
 'football': 35,
 'yes': 36,
 'travelled': 37}

In [32]:
type(tok.word_index)

dict

**So every word here has been given its own unique indexes**

In [33]:
train_story = []
train_question = []
train_answers = []

for story,question,answer in traindata:
    train_story.append(story)
    train_question.append(question)
    train_answers.append(answer)

In [34]:
train_story_seq = tok.texts_to_sequences(train_story)

In [35]:
train_story

[['Mary',
  'moved',
  'to',
  'the',
  'bathroom',
  '.',
  'Sandra',
  'journeyed',
  'to',
  'the',
  'bedroom',
  '.'],
 ['Mary',
  'moved',
  'to',
  'the',
  'bathroom',
  '.',
  'Sandra',
  'journeyed',
  'to',
  'the',
  'bedroom',
  '.',
  'Mary',
  'went',
  'back',
  'to',
  'the',
  'bedroom',
  '.',
  'Daniel',
  'went',
  'back',
  'to',
  'the',
  'hallway',
  '.'],
 ['Mary',
  'moved',
  'to',
  'the',
  'bathroom',
  '.',
  'Sandra',
  'journeyed',
  'to',
  'the',
  'bedroom',
  '.',
  'Mary',
  'went',
  'back',
  'to',
  'the',
  'bedroom',
  '.',
  'Daniel',
  'went',
  'back',
  'to',
  'the',
  'hallway',
  '.',
  'Sandra',
  'went',
  'to',
  'the',
  'kitchen',
  '.',
  'Daniel',
  'went',
  'back',
  'to',
  'the',
  'bathroom',
  '.'],
 ['Mary',
  'moved',
  'to',
  'the',
  'bathroom',
  '.',
  'Sandra',
  'journeyed',
  'to',
  'the',
  'bedroom',
  '.',
  'Mary',
  'went',
  'back',
  'to',
  'the',
  'bedroom',
  '.',
  'Daniel',
  'went',
  'back',
  'to

In [36]:
train_story_seq 

[[34, 1, 12, 7, 11, 4, 5, 25, 12, 7, 30, 4],
 [34,
  1,
  12,
  7,
  11,
  4,
  5,
  25,
  12,
  7,
  30,
  4,
  34,
  16,
  24,
  12,
  7,
  30,
  4,
  6,
  16,
  24,
  12,
  7,
  33,
  4],
 [34,
  1,
  12,
  7,
  11,
  4,
  5,
  25,
  12,
  7,
  30,
  4,
  34,
  16,
  24,
  12,
  7,
  30,
  4,
  6,
  16,
  24,
  12,
  7,
  33,
  4,
  5,
  16,
  12,
  7,
  31,
  4,
  6,
  16,
  24,
  12,
  7,
  11,
  4],
 [34,
  1,
  12,
  7,
  11,
  4,
  5,
  25,
  12,
  7,
  30,
  4,
  34,
  16,
  24,
  12,
  7,
  30,
  4,
  6,
  16,
  24,
  12,
  7,
  33,
  4,
  5,
  16,
  12,
  7,
  31,
  4,
  6,
  16,
  24,
  12,
  7,
  11,
  4,
  6,
  28,
  10,
  7,
  35,
  26,
  4,
  6,
  16,
  12,
  7,
  30,
  4],
 [34,
  1,
  12,
  7,
  11,
  4,
  5,
  25,
  12,
  7,
  30,
  4,
  34,
  16,
  24,
  12,
  7,
  30,
  4,
  6,
  16,
  24,
  12,
  7,
  33,
  4,
  5,
  16,
  12,
  7,
  31,
  4,
  6,
  16,
  24,
  12,
  7,
  11,
  4,
  6,
  28,
  10,
  7,
  35,
  26,
  4,
  6,
  16,
  12,
  7,
  30,
  4,
  13,
  37,


**You can see we got the indexes in the sequence the data is present to us and if you want to check the if indexes are<br>
matching to the correct words then you can check from above where we used `tok.word_index` for printing**

In [37]:
def vectorize_stories(data,word_index=tok.word_index,max_story_len=max_story_len,max_ques_len=max_ques_len):
    S = []  #S is for stories
    Q = []  #Q is for questions
    A = []   #A is for answers
    
    for story, question, answer in data:
        x = [word_index[word.lower()] for word in story]
        xq = [word_index[word.lower()] for word in question]
        y = np.zeros(len(word_index) + 1)    # Index 0 is reserved so we're going to use + 1
        y[word_index[answer]] = 1
        S.append(x)
        Q.append(xq)
        A.append(y)
        
    return (pad_sequences(S, maxlen=max_story_len),pad_sequences(Q, maxlen=max_ques_len), np.array(A))

In [38]:
inputs_train, queries_train, answers_train = vectorize_stories(traindata)

In [39]:
inputs_test, queries_test, answers_test = vectorize_stories(testdata)

In [40]:
answers_test

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 1., 0.]])

In [41]:
tok.word_index['yes']

36

In [42]:
tok.word_index['no']

18

In [43]:
sum(answers_test)

array([  0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.,   0.,   0.,   0.,   0.,   0.,   0., 503.,   0.,   0.,   0.,
         0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.,   0.,   0., 497.,   0.])

**So `yes` is present `497` times at index 5 and `no` is present `503` times at index 34**

The same goes for other data as well which we are working on

### <div class='alert alert-info'> Creating model </font>

In [44]:
from keras.models import Sequential, Model
from tensorflow.keras.layers import Embedding
from keras.layers import Input, Activation, Dense, Permute, Dropout
from keras.layers import add, dot, concatenate
from keras.layers import LSTM

In [45]:
vocab_len

38

**Placeholders for inputs**

We have 2 inputs stories and questions. We use placeholders for this. **Input()** is used to instantiate a Keras tensor.

In [46]:
input_sequence = Input((max_story_len,))
question = Input((max_ques_len,))

## <font color='red'>Lets build the network now</font>

**Input Encoder m**<br>
Input gets embedded to a sequence of vectors

In [47]:
input_encoder_m = Sequential()
input_encoder_m.add(Embedding(input_dim=vocab_len,output_dim=64))
input_encoder_m.add(Dropout(0.3))

**Input Encoder c**<br>
embed the input into a sequence of vectors of size query_maxlen

In [48]:
input_encoder_c = Sequential()
input_encoder_c.add(Embedding(input_dim=vocab_len,output_dim=max_ques_len))
input_encoder_c.add(Dropout(0.3))

**Question Encoder**<br>
embed the question into a sequence of vectors

In [49]:
question_encoder = Sequential()
question_encoder.add(Embedding(input_dim=vocab_len,
                               output_dim=64,
                               input_length=max_ques_len))
question_encoder.add(Dropout(0.3))

**Lets encode the question and sequences**<br>


In [50]:
input_encoded_m = input_encoder_m(input_sequence)
input_encoded_c = input_encoder_c(input_sequence)
question_encoded = question_encoder(question)

**We use the dot product to compute the match between first input vec seq and the query**

In [51]:
match = dot([input_encoded_m, question_encoded], axes=(2, 2))
match = Activation('softmax')(match)

In [52]:
response = add([match, input_encoded_c])  
response = Permute((2, 1))(response) 

**Now we  concatenate the match matrix with the question vector sequence**

In [53]:
answer = concatenate([response, question_encoded])

In [54]:
answer

<KerasTensor: shape=(None, 6, 220) dtype=float32 (created by layer 'concatenate')>

**We are going to reduce the RNN(LSTM)**

In [55]:
answer = LSTM(32)(answer) 

In [56]:
answer = Dropout(0.5)(answer)
answer = Dense(vocab_len)(answer) 

In [57]:
answer = Activation('softmax')(answer)
model = Model([input_sequence, question], answer)
model.compile(optimizer='rmsprop', loss='categorical_crossentropy',
              metrics=['accuracy'])

In [58]:
model.summary()

Model: "model"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_1 (InputLayer)           [(None, 156)]        0           []                               
                                                                                                  
 input_2 (InputLayer)           [(None, 6)]          0           []                               
                                                                                                  
 sequential (Sequential)        (None, None, 64)     2432        ['input_1[0][0]']                
                                                                                                  
 sequential_2 (Sequential)      (None, 6, 64)        2432        ['input_2[0][0]']                
                                                                                              

In [60]:
history = model.fit([inputs_train, queries_train], answers_train,batch_size=32,epochs=150,validation_data=([inputs_test, queries_test], answers_test))

Epoch 1/150
313/313 [==============================] - 3s 10ms/step - loss: 0.2342 - accuracy: 0.9025 - val_loss: 0.4882 - val_accuracy: 0.8260
Epoch 2/150
313/313 [==============================] - 3s 10ms/step - loss: 0.2496 - accuracy: 0.8963 - val_loss: 0.4568 - val_accuracy: 0.8230
Epoch 3/150
313/313 [==============================] - 3s 11ms/step - loss: 0.2375 - accuracy: 0.9022 - val_loss: 0.4956 - val_accuracy: 0.8260
Epoch 4/150
313/313 [==============================] - 3s 10ms/step - loss: 0.2420 - accuracy: 0.9010 - val_loss: 0.4490 - val_accuracy: 0.8300
Epoch 5/150
313/313 [==============================] - 3s 9ms/step - loss: 0.2375 - accuracy: 0.9017 - val_loss: 0.4838 - val_accuracy: 0.8200
Epoch 6/150
313/313 [==============================] - 3s 9ms/step - loss: 0.2347 - accuracy: 0.9056 - val_loss: 0.4942 - val_accuracy: 0.8200
Epoch 7/150
313/313 [==============================] - 3s 9ms/step - loss: 0.2395 - accuracy: 0.9031 - val_loss: 0.4863 - val_accuracy: 0.

Epoch 58/150
313/313 [==============================] - 3s 10ms/step - loss: 0.2094 - accuracy: 0.9136 - val_loss: 0.5200 - val_accuracy: 0.8280
Epoch 59/150
313/313 [==============================] - 4s 11ms/step - loss: 0.2039 - accuracy: 0.9156 - val_loss: 0.5435 - val_accuracy: 0.8130
Epoch 60/150
313/313 [==============================] - 4s 13ms/step - loss: 0.2050 - accuracy: 0.9154 - val_loss: 0.5138 - val_accuracy: 0.8220
Epoch 61/150
313/313 [==============================] - 4s 12ms/step - loss: 0.2101 - accuracy: 0.9115 - val_loss: 0.5150 - val_accuracy: 0.8140
Epoch 62/150
313/313 [==============================] - 3s 10ms/step - loss: 0.2105 - accuracy: 0.9112 - val_loss: 0.4985 - val_accuracy: 0.8270
Epoch 63/150
313/313 [==============================] - 3s 10ms/step - loss: 0.2121 - accuracy: 0.9137 - val_loss: 0.5351 - val_accuracy: 0.8180
Epoch 64/150
313/313 [==============================] - 3s 10ms/step - loss: 0.2128 - accuracy: 0.9160 - val_loss: 0.4873 - val_ac

313/313 [==============================] - 3s 9ms/step - loss: 0.1900 - accuracy: 0.9259 - val_loss: 0.6360 - val_accuracy: 0.8160
Epoch 115/150
313/313 [==============================] - 3s 9ms/step - loss: 0.2000 - accuracy: 0.9226 - val_loss: 0.5353 - val_accuracy: 0.8190
Epoch 116/150
313/313 [==============================] - 3s 9ms/step - loss: 0.2064 - accuracy: 0.9204 - val_loss: 0.5058 - val_accuracy: 0.8340
Epoch 117/150
313/313 [==============================] - 3s 10ms/step - loss: 0.1862 - accuracy: 0.9246 - val_loss: 0.5730 - val_accuracy: 0.8210
Epoch 118/150
313/313 [==============================] - 3s 9ms/step - loss: 0.2010 - accuracy: 0.9219 - val_loss: 0.5445 - val_accuracy: 0.8240
Epoch 119/150
313/313 [==============================] - 3s 9ms/step - loss: 0.1872 - accuracy: 0.9216 - val_loss: 0.5508 - val_accuracy: 0.8260
Epoch 120/150
313/313 [==============================] - 3s 10ms/step - loss: 0.1893 - accuracy: 0.9235 - val_loss: 0.5194 - val_accuracy: 0.82

### Lets save the model now

In [61]:
filename = 'chatbot_epoch_150'
model.save(filename)

INFO:tensorflow:Assets written to: chatbot_epoch_150\assets


INFO:tensorflow:Assets written to: chatbot_epoch_150\assets


## Evaluating the model

In [62]:
model.load_weights(filename)
pred_results = model.predict(([inputs_test, queries_test]))

32/32 [==============================] - 0s 3ms/step


In [64]:
testdata[0][0]

['Mary',
 'got',
 'the',
 'milk',
 'there',
 '.',
 'John',
 'moved',
 'to',
 'the',
 'bedroom',
 '.']

This is the story

In [65]:
testdata[0][1]

['Is', 'John', 'in', 'the', 'kitchen', '?']

This would be the question

In [66]:
testdata[0][2]

'no'

And this should be the answer we should get

In [68]:
story =' '.join(word for word in testdata[0][0])
print(story)

Mary got the milk there . John moved to the bedroom .


In [70]:
query = ' '.join(word for word in testdata[0][1])
print(query)

Is John in the kitchen ?


In [72]:
print("Actual Answer from Data is:",testdata[0][2])

Actual Answer from Data is: no


In [73]:
pred_results

array([[1.1995535e-12, 1.2556545e-12, 1.3593759e-12, ..., 1.2491309e-12,
        4.7522721e-05, 1.1808502e-12],
       [1.8965808e-11, 1.8860401e-11, 2.0287935e-11, ..., 1.8783934e-11,
        2.0675932e-03, 1.8480197e-11],
       [1.3668758e-12, 1.2110241e-12, 1.3346444e-12, ..., 1.3009329e-12,
        8.0092555e-01, 1.1462861e-12],
       ...,
       [1.8916710e-10, 2.0363219e-10, 1.9816375e-10, ..., 1.9611525e-10,
        7.8036515e-03, 1.9844933e-10],
       [4.6302767e-11, 4.3842336e-11, 4.6679813e-11, ..., 4.4853274e-11,
        6.9462121e-01, 4.3603829e-11],
       [2.9775852e-16, 2.2881902e-16, 2.5723967e-16, ..., 2.5706116e-16,
        9.9855059e-01, 2.1512210e-16]], dtype=float32)

**Generating the prediction from the model**

np.argmax- returns indices of the max element of the array in a particular axis.

In [75]:
maxx = np.argmax(pred_results[0])

for key, val in tok.word_index.items():
    if val == maxx:
        k = key

print("Predicted answer is: ", k)
print("Accuracy of the prediction is : ", pred_results[0][maxx])

Predicted answer is:  no
Accuracy of the prediction is :  0.99995244


# <div class='alert alert-success'> Writing our own stories </div>

In [76]:
vocab

{'.',
 '?',
 'Daniel',
 'Is',
 'John',
 'Mary',
 'Sandra',
 'apple',
 'back',
 'bathroom',
 'bedroom',
 'discarded',
 'down',
 'dropped',
 'football',
 'garden',
 'got',
 'grabbed',
 'hallway',
 'in',
 'journeyed',
 'kitchen',
 'left',
 'milk',
 'moved',
 'no',
 'office',
 'picked',
 'put',
 'the',
 'there',
 'to',
 'took',
 'travelled',
 'up',
 'went',
 'yes'}

We have to use words only from the existing vocab

In [95]:
my_story = "Daniel is in the bathroom"
my_story.split()

['Daniel', 'is', 'in', 'the', 'bathroom']

In [96]:
my_question = "Is Daniel in the garden ?"

In [97]:
my_question.split()

['Is', 'Daniel', 'in', 'the', 'garden', '?']

In [98]:
mydata = [(my_story.split(),my_question.split(),'no')]

In [99]:
my_story,my_ques,my_ans = vectorize_stories(mydata)

In [100]:
pred_results = model.predict(([ my_story, my_ques]))

1/1 [==============================] - 0s 17ms/step


In [101]:

maxx = np.argmax(pred_results[0])

for key, val in tok.word_index.items():
    if val == maxx:
        k = key

print("Predicted answer is: ", k)
print("Probability of certainty was: ", pred_results[0][maxx])

Predicted answer is:  no
Probability of certainty was:  0.99798393


# <font color='green'>There you go</font>